trinity/config.py

```diff
@@ -440,14 +444,77 @@ def nodekey(self, value: Union[bytes, PrivateKey]) -> None:
                 f"`PrivateKey` instance: got {type(self._nodekey)}"
             )
 
+    @contextmanager
+    def process_id_file(self, process_name: str):  # type: ignore
+        with PidFile(process_name, self.data_dir):
+            yield
+
     @classmethod
-    def from_parser_args(cls, parser_args: argparse.Namespace) -> 'TrinityConfig':
+    def from_parser_args(cls,
+                         parser_args: argparse.Namespace,
+                         app_config_types: Iterable[Type['BaseAppConfig']]) -> 'TrinityConfig':
         """
         Helper function for initializing from the namespace object produced by
         an ``argparse.ArgumentParser``
         """
         constructor_kwargs = construct_trinity_config_params(parser_args)
-        return cls(**constructor_kwargs)
+        trinity_config = cls(**constructor_kwargs)
+
+        trinity_config.initialize_app_configs(parser_args, app_config_types)
+
+        return trinity_config
+
+    def initialize_app_configs(self,
+                               parser_args: argparse.Namespace,
+                               app_config_types: Iterable[Type['BaseAppConfig']]) -> None:
+
+        for app_config_type in app_config_types:
+            self.add_app_config(app_config_type.from_parser_args(parser_args, self))
+
+    def add_app_config(self, app_config: 'BaseAppConfig') -> None:
+        self._app_configs[type(app_config)] = app_config
+
+    def has_app_config(self, app_config_type: Type['BaseAppConfig']) -> bool:
+        return app_config_type in self._app_configs.keys()
+
+    def get_app_config(self, app_config_type: Type[TAppConfig]) -> TAppConfig:
+        # We want this API to return the specific type of the app config that is requested.
+        # Our backing field only knows that it is holding `BaseAppConfig`'s but not concrete types
+        return cast(TAppConfig, self._app_configs[app_config_type])
+
+
+class BaseAppConfig(ABC):
+
+    @classmethod
+    @abstractmethod
+    def from_parser_args(cls,
+                         args: argparse.Namespace,
+                         base_config: TrinityConfig) -> 'BaseAppConfig':
+        pass
+
+
+class DefaultBaseAppConfig(BaseAppConfig):
+
+    def __init__(self, args: argparse.Namespace, base_config: TrinityConfig):
+        self.args = args
+        self.base_config = base_config
+
+    @classmethod
+    def from_parser_args(cls,
+                         args: argparse.Namespace,
+                         base_config: TrinityConfig) -> 'BaseAppConfig':
+        return cls(args, base_config)
+
+
+class Eth1AppConfig(DefaultBaseAppConfig):
+
+    @property
+    def is_light_mode(self) -> bool:
+        return self.base_config.sync_mode == SYNC_LIGHT
+
+    @property
+    def is_full_mode(self) -> bool:
+        return self.base_config.sync_mode == SYNC_FULL
```

Comment:
@pipermerriam One could argue that `sync_mode` should move into the `Eth1AppConfig` as well but the problem is that this spreads like wild fire. Because the `database_dir` depends on the `sync_mode` but I don't think we want to move the `database_dir` as it really is a core thing that will be needed for any client that trinity would support.

So, my current syncing is, let `sync_mode` continue to be a `str` in `TrinityConfig` but leave it up to the specific app config to interpret that string (there's currently a hard validation against `full` / `light` but it we could make that more flexible)

The bigger problem really is that everything in our current scheme is modeled around the idea that we are always talking about an Ethereum 1 implementation.

```
trinity/<network>/chain/<sync-mode>
trinity/<network>/logs
...
```

I guess we don't want to abuse the network identifier for the purpose of supporting the beacon node ;)

So, it seems to boil down to two options:

1. Add another top level hierarchy `trinity/<client|protocol|whatever>/<network>/chain/<sync-mode>`

2. Use an entirely different `trinity_root_dir` so that it becomes `trinity-beacon/<network>/chain/<sync-mode>`

I'm in favor of the second option for multiple reasons:

A.) It is backwards compatible. The current trinity does not need to migrate its dirs

B.) It feels natural **if** we continue to treat the beacon node as an entirely different executable (it currently is `trinity-beacon`.

I might go ahead and give B a shot to see where that falls apart :)

Reply:
Went ahead with that idea and added a commit on top https://github.com/ethereum/py-evm/pull/1566/commits/2df39dbfd6eddcb74b8d330c9651d57fb6f0e63c

To recap:

- we use a different root path per application
- the `TrinityConfig` only holds what is anticipated to be used across all applications (`SYNC_MODE` is debatable but since it influences very low level things like `database_dir` I feel we should just keep that a core config) 
- everything else goes into additionally registered sub configs

Reply:
What about:

- `trinity/mainnet/eth1/<sync-mode>`
- `trinity/mainnet/beacon/...'
- 'trinity/mainnet/shard-01/...'

It seems beneficial to co-locate all of the various *chains* under a single root directory for the network they are part of.

Reply:
Regarding `database_dir`.

I think this is actually an `eth1` specific thing and **should-not** be shared by any of the other applications (beacon/shard), as each of them should have their own databases.

Reply:
>or the network they are part of.

Can we really say that the beacon node and the Ethereum are in the same network? All the different shards that a beacon node would download could sure all live under one `trinity-beacon/mainnet/shard-0|shard-1|...` space but since shards do not exist in Ethereum 1 it feels natural to me to just use `/<network>/<beacon|shard-n>` as a new schema for the beacon node.

I think the two primary reasons why I'd be in favor of a different `<root-dir>` rather than changing the existing schema are:

1. It currently looks like `trinity-beacon` will be a different executable, a different program that just builds on a shared code base. So in that sense, a different root dir makes most sense to me.

2. Changing the existing schema is a hard breaking change (sure we could detect and migrate existing data directories but still)

Reply:
I will move this into another PR so that we can discuss this separately. I'd also be interested to hear what @carver thinks about it. Something in me really likes to treat the beacon node as a separate program here. E.g what about the `NodeKey` would the regular trinity node share that with the beacon node? It seems cleaner to me to keep the root directories of each of them clearly separated..also in terms of logs etc.

Reply:
> Plus, isn't the litmus test test that if we ever would decide to move the beacon client into its entirely own project (instead of just maintaining a separate entry point / bootstrapping in the same code base) then we would for sure use a different root directory anyway? So, this would make a potential migrating easier long term by treating them as two different programs right from the start.

This seems like a fairly strong argument. If it runs as a separate project, maybe it belongs in its own root dir.

OTOH: *maybe* `trinity` becomes the all-in-one command, and what we currently call `trinity` moves to `trinity-eth1` or something. Then implicitly `trinity` runs both `trinity-beacon` and `trinity-eth1`. So maybe the directory structure is something like `trinity/beacon-mainnet/...` and `trinity/eth1-mainnet/...`.

Can you link the other issue/PR to move the discussion to?

Reply:
Yep, it's in this PR (that proposes a solution) https://github.com/ethereum/py-evm/pull/1603